### PR TITLE
New tooling

### PR DESCRIPTION
**Pull request by jvanbruegge**
This PR is to simplify the tooling for snabbdom. Roadmap for this PR:
- [x] Drop webpack
- [x] Drop custom typescript compiler plugin
- [x] Implement #913 (needed for simpler tooling, especially karma)
- [x] Offer cjs and esm bundles for easier use with nodejs
- [x] Just use karma for typescript tests
- [x] Add prettier
- [x] Port over jsx tests
- [ ] ~~Port over benchmark tests~~

### CODE REVIEW

**Comment by jvanbruegge** [comment]
The CI won't pass, but the karma tests for all *.ts (not tsx) pass already locally

**Comment by mightyiam** [comment]
@jvanbruegge you can approve deploys now.

File: karma.conf.js
Comment (mightyiam) [inline karma.conf.js:23]: I don't like karma-typescript or most of these kinds of creations. I see them as unnecessary couplings. I like chaining my tools myself. It seems simpler to me and with less of the pipeline under the hood of some dependency.

**Review by mightyiam — COMMENTED** [review]
Wow, this is a big change. It's like... a lot of my work is going down the drain. I don't agree with most of this. And this PR is a bit too large to review comfortably for me. But if you'd like to take the reigns on this project then...

**Comment by jvanbruegge** [comment]
To make contributions easier and ease maintenance, I want to get rid of as much custom tooling as possible and replace it with standard tools. The main reason is that custom tooling has to be maintained by us, while more eyes are on the standard stuff.

For me, ideally we would just have `karma`, `typescript`, `prettier` and `rollup` and nothing else.

If you want to look more towards rust and only do small maintenance (which I completely understand), then yes, I'd like to take over the reigns, if you are ok with that. I won't start making major, publicly visible decisions on my own, but consult and discuss with the rest of the maintainers, be it you or @mreinstein. If this is fine by you, could you make me an owner on the snabbdom repo? I want to adjust CI a bit, so we can run a stripped down version (no browserstack, thus no secrets) even without approval

**Comment by mreinstein** [comment]
I don't know if I'm alone in this feeling but similar to typescript, I find the output of `prettier` to be ironically very un-pretty. 

While I understand the herd's movement towards these automated tools, I do lament the decrease in readability they incur.

**Comment by mreinstein** [comment]
>  It's like... a lot of my work is going down the drain.

@mightyiam  I would encourage you to re-think your perspective on this.  I don't think any of the maintenance stuff you've done has gone down the drain, it's just an ongoing task, and things inevitably change. The ecosystem has changed and it's more of an evolution than any kind of deletion.

Something separate to consider is that commonjs is on the way out. Every production version of node currently supports es modules except for node v10, and that loses maintenance support in April (see https://nodejs.org/en/about/releases/)

So while I don't think we should outright break commonjs support today, the writing is on the wall, it's on the way out. We should operate under the assumption that we need to support cjs today but that it's a temporary thing, and maybe a year from now it might be ok to just drop it (depending on how fully the ecosystem embraces it.)

**Comment by mreinstein** [comment]
2 concerns:

1-  There is no included documentation around the changes to how one`require` / `import` s snabbdom. That should be included with any changes to that stuff, so the commit is atomic.

2- It's unclear to me why we need to run `rollup` for the es module build. Can't we just point at the es module directly, and only use `rollup` to produce a temporary `cjs` build? The nice thing about this is if/when commonjs becomes unneeded we can just drop rollup.

thoughts?

**Comment by wclr** [comment]
> I find the output of prettier to be ironically very un-pretty.

This is just about standardization of the code convenience for developers when you should not care about formatting at all. For self-sufficient small lib, it really doesn't matter what auto-formatting is used, the thing that it should be used not to create additional work and hassle.

**Comment by mreinstein** [comment]
> For self-sufficient small lib, it really doesn't matter what auto-formatting is used

I agree with this statement, in isolation. It is a small library, and given how critical it is, making it readable by humans is really, really high priority in my opinion.

While DOM diff/patching is not rocket science it is critically important and underpins several high value frameworks (vue, cyclejs, snabby.)

**Comment by wclr** [comment]
> making it readable by humans is really, really high priority in my opinion

the prettier output is absolutly readable even with semicolons 😄

**Comment by mreinstein** [comment]
> the prettier output is absolutly readable even with semicolons

Semicolons are a trivial non-issue. The horrific way it indents and spaces lines in some cases is of more concern.

**Comment by jvanbruegge** [comment]
> There is no included documentation around the changes to how onerequire / import s snabbdom

Updated the README now in the same commit

>  Can't we just point at the es module directly, and only use rollup to produce a temporary cjs build?

Did that now too

File: README.md
Comment (mreinstein) [inline README.md:155]: In the interests of making all of the example code sections smaller and more readable, I think we can omit all of the `import` statements _after_ the first example. We don't need to demo the import process in every example snippet.

File: src/index.ts
Comment (mreinstein) [inline src/index.ts:28]: let's kill this thing immediately. Will make the module a little smaller and also resolve https://github.com/snabbdom/snabbdom/issues/517

Comment (jvanbruegge) [reply src/index.ts:28]: Wanted to do that in a seperate PR

Comment (mreinstein) [reply src/index.ts:28]: I guess that's fine, since we're not actually pushing this out to a new npm version anyway. Just figured it's a little silly to carry forward this piece we know we don't want. but it's not a big deal.

**Comment by jvanbruegge** [comment]
So, eslint is back at it, now together with prettier. Next step for me is fixing the JSX stuff

**Comment by mightyiam** [comment]
Could we do small PRs, instead of this?

**Comment by mightyiam** [comment]
Seriously. Is the tooling that I perceive as superbly set up so aweful that they must be replaced in one fell swoop?

**Comment by mreinstein** [comment]
@jvanbruegge thank you for pushing these changes forward! I don't agree with everything but I'm glad to see the project evolve.

**Comment by kuraga** [comment]
Drop karma, too would be brilliant :)

**Comment by kuraga** [comment]
> Drop karma, too would be brilliant :)

BTW, I meant to use something lightweightweight like [tman](https://github.com/thunks/tman).

**Comment by jvanbruegge** [comment]
karma != mocha
karma allows us to run the tests in the browser fairly easily, even with Browserstack

**Comment by jvanbruegge** [comment]
> Could we do small PRs, instead of this?

I don't want to leave the master in a broken state. IMO the commits are cleanly separated regardless if they sit in 4 or 1 PR

**Comment by kuraga** [comment]
> karma != mocha
> karma allows us to run the tests in the browser fairly easily, even with Browserstack

Tman too, but not so easy. But lightweight :)

**Comment by paldepind** [comment]
Thank you for working on this @jvanbruegge.

> To make contributions easier and ease maintenance, I want to get rid of as much custom tooling as possible and replace it with standard tools. The main reason is that custom tooling has to be maintained by us, while more eyes are on the standard stuff.

I agree with this sentiment. If we can get an equally good result for end users but while mainting fewer lines of configuration then that seems like a net benefit to me.

I agree with @mightyiam that this PR is a bit difficult to review given the number of things changed. It also makes it harder for people to voice their opinions on individual things. But, I also understand that doing this at once is probably easier for you @jvanbruegge. If it's not a big problem I think it would be nice to split this up a bit, but alternatively you could maybe just summarise briefly the various changes?

> Seriously. Is the tooling that I perceive as superbly set up so aweful that they must be replaced in one fell swoop?

I don't think anyone considers the current tooling to be awful 🙂. I think your should share your opinion on the changes supported with technical arguments, then we can all take them into consideration and decide on what is best going forward.

**Comment by paldepind** [comment]
I think this seems very reasonable overall. And most of it does seem to simplify the tooling.

The only part of the PR that I'm a bit worried about is the "export everything from 'snabbdom'" part. I've added a comment regarding that over in the related issue https://github.com/snabbdom/snabbdom/issues/913#issuecomment-786480470.

**Comment by kuraga** [comment]
> karma != mocha
> karma allows us to run the tests in the browser fairly easily, even with Browserstack

One more comment about Tman. I know we don't need it here (as tooling works already) but for your knowledge.

It's brilliant because it's lightweight (as such as Snabbdom :smile: ). It's not mocha, it's just (only!) test runner.

To run tests in browser I [just use Rollup](https://github.com/kuraga/rotorjs/blob/master/scripts/testBrowser.js) without big issues (but yes sometimes they exist).

**Comment by wclr** [comment]
I'm on board with this. The tooling that is currently used for tests etc doesn't seem to be well-tuned, some times tests fail locally. And overall config seemed to be a little bit complicated.

**Comment by paldepind** [comment]
@wclr

> Easy to ask when you are not performing the task.

Please stop the aggressive attitude and be nice to everybody here. I will not tolerate comments like that and this is not the first one (https://github.com/snabbdom/snabbdom/issues/932#issuecomment-779917547). @mightyiam has made tremendous contributions to Snabbdom and was making a perfectly reasonable comment.

Everybody, please keep the comments constructive and on-topic. Thank you 🙏

**Comment by paldepind** [comment]
To be clear. I think we (and by "we" let's acknowledge that @jvanbruegge is the one doing the work) should absolutely continue with this PR. It brings a bunch of great changes with it.

I think we should do a release after this PR is merged. Exporting everything from the main module is going to be a small breaking change, but it will also simplify the whole import process (thanks to tree-shaking) and fix issues such as #723.

**Comment by jvanbruegge** [comment]
Yeah, I will continue, I just have an exam today and on Friday, so that kind of hogged my free time a bit 😅

**Comment by paldepind** [comment]
@jvanbruegge No problem. It was not to rush you in any way 😄. My comment was to make sure that we all agreed on that and to make sure you didn't feel discouraged.

Good luck with your exam 🤞

**Comment by jvanbruegge** [comment]
I finally found out how typescript wants you to JSX typings. Now there is no global declarations file, but instead it is added to the factory function itself.

**Comment by jvanbruegge** [comment]
The benchmark tests test the current version against the last released snabbdom version. For now I would not port those over, until we have release v3.0 because there are two issues:
- The way snabbdom exports stuff, makes it hard to get it to work with karma
- After releasing v3 all the workarounds would be useless and I don't really want to waste a lot of time there now
Is that ok?

**Comment by paldepind** [comment]
@jvanbruegge 

> The benchmark tests test the current version against the last released snabbdom version. For now I would not port those over, until we have release v3.0 because there are two issues: [...]

Yes, that seems very reasonable to me.

**Comment by jvanbruegge** [comment]
I've swapped `commitlint` for my new package `commithelper`. The main reason is that commitlint can't provide the user with a convenient prompt to create a fitting commit message, while commitizen is not able to lint an existing commit message. commithelper does both.

**Comment by jvanbruegge** [comment]
I've rebased and reworded all commit messages to match the linting. This includes adding explainatory bodies and mentioning issues closed and breaking changes

**Comment by jvanbruegge** [comment]
@paldepind Can you make me owner of the snabbdom repo? Always having to approve CI is annoying

**Comment by jvanbruegge** [comment]
So, this would be done from my side except for CI which I can't do until I have owner access to this repo. Once version 3.0 is released I will tackle the benchmark tests as explained earlier.
Some statistics of this PR:
|                      | master | new tooling | difference |
|----------------------|--------|-------------|------------|
| config files         | 12     | 10          | -2         |
| direct dependencies  | 52     | 40          | -12        |
| overall dependencies | 1483   | 1069        | -414       |

**Comment by jvanbruegge** [comment]
ping @paldepind

**Comment by mreinstein** [comment]
@jvanbruegge great work, man! Thanks for doing this!

**Comment by paldepind** [comment]
@jvanbruegge Sorry for the delay. I've made you an owner of the Snabbdom organization and sent you an invitation to become a maintainer of the Snabbdom package on npm. On a related note, do you know if the ["cycle" npm account](https://www.npmjs.com/~cycle) is used for anything? Or can we remove that account from the npm package?

PR looks great to me! 👍

**Comment by jvanbruegge** [comment]
Thanks!
I am not aware that the cycle account is used, it should be save to remove it

**Comment by paldepind** [comment]
> I am not aware that the cycle account is used, it should be save to remove it

Thanks, great. I've removed it.

**Comment by jvanbruegge** [comment]
So, CI works, runs on every commit for every PR now. I was also able to reactivate old browsers in the browserstack config. @paldepind @mreinstein please do a final review

Just a side note, this PR closes 8 open issues :tada:

**Comment by jvanbruegge** [comment]
Once this is merged I can release version 3.0

File: package.json
Comment (mreinstein) [inline package.json:5]: I'm a little sad to see this. `type: "module"` is the node sanctioned way of declaring this module to be es based. Going back to cjs feels a bit regressive. :(

**Review by mreinstein — APPROVED** [review]

File: package.json
Comment (jvanbruegge) [reply package.json:5]: The problem is AFAICT that node will never pick the entry from `module`, always `main`. So if we set `type: module` we would need to throw out cjs completely. And at least I have often issues with testing and ES modules

**Comment by paldepind** [comment]
Thanks a lot for doing this work @jvanbruegge.

**Comment by jvanbruegge** [comment]
@paldepind would it be possible to undo the merge and do a merge commit? having one gigantic commit on master that touches almost 100 files is not really helpful. Of you want I can also do it

**Comment by jvanbruegge** [comment]
Oh, I also see that this totally breaks the changelog because the commit messages do not have subject fields any more. I will change this now

**Comment by kuraga** [comment]
🥳

**Pull request merged**
